app/src/server/tasks/generateInvoices.task.ts

```diff
@@ -29,123 +51,122 @@ export const generateInvoices = defineTask({
       })
-      .returning(["id", "createdAt", "slug"])
+      .returning([
+        "id",
+        "createdAt",
+        "slug",
+        "amount",
+        "billingPeriod",
+        "status",
+        "description",
+        "projectId",
+      ])
       .executeTakeFirst();
 
+    if (!invoice) throw new Error("No data available for invoice generation");
+
     // 2. Associate stats with invoice
-    if (invoice) {
+    await tx
+      .updateTable("UsageLog")
+      .set({
+        invoiceId: invoice.id,
+      })
+      .where("projectId", "=", projectId)
+      .where("createdAt", ">=", startDate)
+      .where("createdAt", "<=", endDate)
+      .execute();
+
+    // 3. Calculate usage stats
+    const stats = await getStats(
+      tx.selectFrom("UsageLog as ul").where("ul.invoiceId", "=", invoice.id),
+    );
+
+    // 4. Calculate available credits
+    const creditsAvailable = await tx
+      .selectFrom("CreditAdjustment as ca")
+      .where("ca.projectId", "=", projectId)
+      .where("ca.createdAt", "<=", endDate)
+      .select(({ fn }) => fn.sum(sql<number>`amount`).as("amount"))
+      .executeTakeFirst();
+
+    if (!stats || !creditsAvailable) {
+      throw new Error("No data available for invoice generation");
+    }
+    // 5. Calculate total spend
+    const { totalSpent, creditsUsed, remainingCredits } = calculateSpendingsWithCredits(
+      Number(stats?.cost ?? 0),
+      Number(creditsAvailable.amount ?? 0),
+    );
+
+    // 6. Update invoice
+    invoice = await tx
+      .updateTable("Invoice")
+      .set({
+        amount: totalSpent,
+        status: totalSpent >= 1 ? "PENDING" : "CANCELLED", // Minimum $1 charge
+        description: JSON.stringify(
+          getInvoiceDescription({
+            totalInferenceSpend: Number(stats?.totalInferenceSpend ?? 0),
+            totalTrainingSpend: Number(stats?.totalTrainingSpend ?? 0),
+            totalInputTokens: Number(stats?.totalInputTokens ?? 0),
+            totalOutputTokens: Number(stats?.totalOutputTokens ?? 0),
+            totalTrainingTokens: Number(stats?.totalTrainingTokens ?? 0),
+            creditsUsed,
+            remainingCredits,
+          }),
+        ),
+      })
+      .where("id", "=", invoice.id)
+      .returning([
+        "id",
+        "createdAt",
+        "slug",
+        "amount",
+        "billingPeriod",
+        "status",
+        "description",
+        "projectId",
+      ])
+      .executeTakeFirst();
+    if (!invoice) throw new Error("No data available for invoice generation");
+
+    // 6. Create negative credit adjustment if credits were used
+    if (invoice && creditsUsed > 0) {
       await tx
-        .updateTable("UsageLog")
-        .set({
+        .insertInto("CreditAdjustment")
+        .values({
+          id: uuidv4(),
+          projectId: projectId,
+          amount: -creditsUsed,
+          //Adjustments should be created on the first day of the next month, when invoices are generated
```

Comment:
Nit: add space between // and the rest of your comment.

Reply:
Oh, I always forget about it!